libs/pika/synchronization/src/detail/condition_variable.cpp

```diff
@@ -99,34 +99,22 @@ namespace pika::detail {
         // swap the list
         queue_type queue;
         queue.swap(queue_);
-        lock.unlock();
 
-        if (!queue.empty())
-        {
-            // update reference to queue for all queue entries
-            for (queue_entry& qe : queue) qe.q_ = &queue;
-
-            do {
-                auto ctx = queue.front().ctx_;
-
-                // remove item from queue before error handling
-                queue.front().ctx_.reset();
-                queue.pop_front();
+        // update reference to queue for all queue entries
+        for (queue_entry& qe : queue) qe.q_ = &queue;
 
-                if (PIKA_UNLIKELY(!ctx))
-                {
-                    lock.lock();
-                    prepend_entries(lock, queue);
-                    lock.unlock();
-
-                    PIKA_THROWS_IF(ec, pika::error::null_thread_id,
-                        "condition_variable::notify_all", "null thread id encountered");
-                    return;
-                }
```

Comment:
To the best of my knowledge, yes. Getting a null thread id here would mean that things have gone badly, badly wrong, and I think an assertion is more appropriate (there is an assertion there now, instead of a `throw`). Basically, the thread should either already have removed itself from the queue if it's woken up in the middle of a notification, or it should still be in the queue with a non-null thread id.